File: Greedy/remove-k-digits.py
from typing import List
# 48ms 58.22%
# 13.2Mb 38.25
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:

        stack = []
        count = 0
        nums = len(num) - k
        for i in range(len(num)):

            current = num[i]
            if stack and len(stack) == nums and num[i] >= stack[-1]:
                count += 1
            else:
                while stack and current < stack[-1] and count < k:
                    stack.pop()
                    count += 1
                stack.append(current)

        while count < k:
            stack.pop()
            count += 1

        while len(stack) >= 2 and stack[0] == '0':
            stack.pop(0)
        if len(stack) == 0:
            stack.append('0')
        return ''.join(stack)

a = Solution()
l = '10'
print(a.removeKdigits(l, 2))

File: Greedy/remove-k-digits.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string removeKdigits(string num, int k) {
        if (k==num.size()) return "0";
        stack<char> st;
        const int p=num.size()-k;
        int num_left=num.size(), index=0;
        while (index<num.size()){
            if (st.empty()) {
                st.push(num[index]);
                index += 1;
                num_left -= 1;
                continue;
            } 
            while (!st.empty() && st.size() + num_left > p && st.top() > num[index]) st.pop();
            if (st.size() < p) st.push(num[index]);
            num_left -= 1;
            index += 1;
        }
        string ans;
        while (!st.empty()){
            char c = st.top();
            st.pop();
            ans += c;
        }
        reverse(ans.begin(),ans.end());
        while (ans.size() > 1 && ans[0] == '0') ans = ans.substr(1);
        return ans;
    }
};

int main(){
    Solution sol;
    string s = "";
    int k =50000;
    cout << sol.removeKdigits(s, k) << endl;
}

